src/nupic/algorithms/BacktrackingTMCpp.hpp

```diff
@@ -0,0 +1,586 @@
+ *
+ * @param verbosity: (int) Controls the verbosity of the TM diagnostic
+ *         output:
+ *                     - verbosity == 0: silent
+ *                     - verbosity in [1..6]: increasing levels of verbosity
+ *
+ * @param pamLength: (int) Number of time steps to remain in "Pay Attention
+ *          Mode" after we detect we've reached the end of a learned sequence.
+ *          Setting this to 0 disables PAM mode. When we are in PAM mode,
+ *          we do not burst unpredicted columns during learning, which in
+ *          turn prevents us from falling into a previously learned sequence
+ *          for a while (until we run through another 'pamLength' steps).
+ *
+ *          The advantage of PAM mode is that it requires fewer
+ *          presentations to learn a set of sequences which share elements. The
+ *          disadvantage of PAM mode is that if a learned sequence is
+ *immediately followed by set set of elements that should be learned as a 2nd
+ *          sequence, the first 'pamLength' elements of that sequence will not
+ *be learned as part of that 2nd sequence.
+ *
+ * @param maxInfBacktrack: (int) How many previous inputs to keep in a
+ *          buffer for inference backtracking.
+ *
+ * @param maxLrnBacktrack: (int) How many previous inputs to keep in a
+ *          buffer for learning backtracking.
+ *
+ * @param maxSeqLength: (int) If not 0, we will never learn more than
+ *          'maxSeqLength' inputs in a row without starting over at
+ *          start cells. This sets an upper bound on the length of
+ *          learned sequences and thus is another means (besides
+ *          'maxAge' and 'globalDecay') by which to limit how much the
+ *          TM tries to learn.
+ *
+ * @param maxSegmentsPerCell: (int) The maximum number of segments allowed
+ *          on a cell. This is used to turn on "fixed size CLA" mode. When
+ *          in effect, 'globalDecay' is not applicable and must be set to 0 and
+ *          'maxAge' must be set to 0. When this is used (> 0),
+ *'maxSynapsesPerSegment' must also be > 0.
+ *
+ * @param maxSynapsesPerSegment: (int) The maximum number of synapses
+ *          allowed in a segment. This is used to turn on "fixed size CLA" mode.
+ *          When in effect, 'globalDecay' is not applicable and must be set
+ *          to 0, and 'maxAge' must be set to 0. When this is used (> 0),
+ *          'maxSegmentsPerCell' must also be > 0.
+ *
+ * @param outputType: (string) Can be one of the following (default 'normal')
+ *            - 'normal': output the OR of the active and predicted state.
+ *            - 'activeState': output only the active state.
+ *            - 'activeState1CellPerCol': output only the active
+ *               state, and at most 1 cell/column. If more than 1 cell
+ *               is active in a column, the one with the highest confidence
+ *               is sent up.
+ *
+ *************/
+
+class BacktrackingTMCpp {
```

Comment:
should extend TempralMemory 
so we can use
`TemporalMemory tm = (Backtracking)TemporalMemory(...)`;